File: Algorithm/baekjoon/cpp/BJ1761_1.cc
#include <cstdio>
#include <algorithm>
#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <sstream>
using namespace std;
struct Edge {
    int to, cost;
    Edge(int to, int cost) : to(to), cost(cost) {
    }
};
const int MAX = 100111;
vector<Edge> a[MAX];
int parent[MAX];
bool check[MAX];
int depth[MAX];
int dist[MAX];
int lca(int u, int v) {
    if (depth[u] < depth[v]) {
        swap(u,v);
    }
    int output = 0;
    while (depth[u] != depth[v]) {
        output += dist[u];
        u = parent[u];
    }
    while (u != v) {
        output += dist[u];
        output += dist[v];
        u = parent[u];
        v = parent[v];
    }
    return output;
}
int main() {
    int n;
    scanf("%d",&n);
    for (int i=0; i<n-1; i++) {
        int u,v,w;
        scanf("%d %d %d",&u,&v,&w);
        a[u].push_back(Edge(v,w));
        a[v].push_back(Edge(u,w));
    }
    depth[1] = 0;
    check[1] = true;
    queue<int> q;
    q.push(1);
    parent[1] = 0;
    while (!q.empty()) {
        int x = q.front();
        q.pop();
        for (Edge &e : a[x]) {
            int y = e.to;
            if (!check[y]) {
                depth[y] = depth[x] + 1;
                dist[y] = e.cost;
                check[y] = true;
                parent[y] = x;
                q.push(y);
            }
        }
    }
    int m;
    scanf("%d",&m);
    while (m--) {
        int u, v;
        scanf("%d %d",&u,&v);
        printf("%d\n",lca(u, v));
    }
    return 0;
}

File: Algorithm/baekjoon/java/BJ1761_1.java
package baekjoon.java;

import java.util.*;

/**
 * Created by changmin on 2018. 1. 5..
 */
public class BJ1761_1 {
    static class Edge {
        public int to;
        public int cost;
        Edge(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }
    }
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        List<Edge>[] a = (List<Edge>[]) new List[n+1];
        for (int i=1; i<=n; i++) {
            a[i] = new ArrayList<Edge>();
        }
        for (int i=1; i<=n-1; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            int z = sc.nextInt();
            a[x].add(new Edge(y,z));
            a[y].add(new Edge(x,z));
        }
        int[] depth = new int[n+1];
        boolean[] check = new boolean[n+1];
        int[] parent = new int[n+1];
        int[] dist = new int[n+1];
        int start = 1;
        Queue<Integer> q = new LinkedList<Integer>();
        check[start] = true;
        depth[start] = 0;
        parent[start] = 0;
        q.add(start);
        while (!q.isEmpty()) {
            int x = q.remove();
            for (Edge e : a[x]) {
                int y = e.to;
                if (check[y] == false) {
                    dist[y] = e.cost;
                    depth[y] = depth[x] + 1;
                    check[y] = true;
                    parent[y] = x;
                    q.add(y);
                }
            }
        }
        int m = sc.nextInt();
        while (m-- > 0) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            int ans = 0;
            if (depth[x] < depth[y]) {
                int temp = x;
                x = y;
                y = temp;
            }
            while (depth[x] != depth[y]) {
                ans += dist[x];
                x = parent[x];
            }
            while (x != y) {
                ans += dist[x] + dist[y];
                x = parent[x];
                y = parent[y];
            }
            System.out.println(ans);
        }
    }
}
